File: problems/5/solution.go
package main

// The difference between uppercase and lowercase ascii characters
var Difference = byte('a') - byte('A')

// IsAlnum returns true if the input byte is alphanumeric
// when converted to an ascii character.
func IsAlpha(ch byte) bool {
	return ch >= byte('A') && ch <= byte('z')
}

func IsNum(ch byte) bool {
	return ch >= byte('0') && ch <= byte('9')
}

// IsAlnum returns true if the input byte is alphanumeric
// when converted to an ascii character.
func IsAlnum(ch byte) bool {
	return IsAlpha(ch) || IsNum(ch)
}

// Returns true if the input bytes represent the same
// character (case insensitive).
func BytesEqual(one byte, other byte) bool {
	if IsNum(one) || IsNum(other) {
		return one == other
	}

	switch one - other {
	case 0, Difference, -Difference:
		// check if they're both alphabetical
		return true
	default:
		return false
	}
}

// ValidPalindrome returns true if the input string
// is a palindrome. Ignores all non-alphanumeric characters
// in the string.  An empty string is considered a palindrome.
func ValidPalindrome(s string) bool {
	if len(s) < 2 {
		return true
	}

	left := 0
	right := len(s) - 1
	for left < right {
		for left < len(s) && !IsAlnum(s[left]) {
			left++
		}
		for right > 0 && !IsAlnum(s[right]) {
			right--
		}

		if left > right {
			return true
		}

		if !BytesEqual(s[left], s[right]) {
			return false
		}

		left++
		right--
	}

	return true
}

File: problems/5/solution.py
def valid_palindrome(s: str) -> bool:
    """
    Returns True if s is a palindrome. Ignores all non-alphanumeric
    characters and is case-insensitive.

    Args:
        s (str): string to be palindrome tested
    Returns:
        bool: True, if the string is a palindrome
    """
    # remove all non-ascii characters and make lowercase
    if len(s) < 2:
        return True
    if len(s) == 2:
        return s[0] == s[1]

    left = 0
    right = len(s) - 1
    while left < right:
        while left < len(s) and not s[left].isalnum():
            left += 1
        while right > 0 and not s[right].isalnum():
            right -= 1

        if left >= right:
            return True

        if s[left].lower() != s[right].lower():
            return False

        left += 1
        right -= 1

    return True


def test_valid_palindrome():
    assert valid_palindrome("A man, a plan, a canal: Panama") == True
    assert valid_palindrome("race a car") == False
    assert valid_palindrome(" ") == True


if __name__ == "__main__":
    import pytest
    pytest.main(["-xv", __file__])
